platforms/core-configuration/declarative-dsl-core/src/main/kotlin/org/gradle/internal/declarativedsl/mappingToJvm/RuntimeFunctionResolver.kt

```diff
@@ -44,44 +47,33 @@ class MemberFunctionResolver(private val configureLambdaHandler: ConfigureLambda
         val parameterBindingStub = schemaFunction.parameters.associateWith { Any() }
         val hasConfigureLambda = (schemaFunction.semantics as? FunctionSemantics.ConfigureSemantics)?.configureBlockRequirement?.allows ?: false
 
+        fun signature(function: KFunction<*>): String =
+            function::class.memberProperties.first { it.name == "signature" }.apply { isAccessible = true }.call(function) as String
+
+        fun matchesAnnotationsRecursively(function: KFunction<*>, receiverClass: KClass<*>, predicate: (Annotation) -> Boolean): Boolean =
+            if (function.annotations.any(predicate)) {
+                true
+            } else {
+                val functionSignature = signature(function)
+                val parentClasses = receiverClass.supertypes.mapNotNull { it.classifier }.filterIsInstance<KClass<*>>()
+                parentClasses.any { parentClass ->
+                    val parentFunction = parentClass.memberFunctions.firstOrNull { signature(it) == functionSignature }
+                    parentFunction?.let { matchesAnnotationsRecursively(it, parentClass, predicate) } ?: false
+                }
+            }
+
         val resolutions = receiverClass.memberFunctions
             .filter { function -> function.name == schemaFunction.simpleName && FunctionBinding.convertBinding(function, Any(), parameterBindingStub, hasConfigureLambda, configureLambdaHandler) != null }
+            .filter { f -> matchesAnnotationsRecursively(f, receiverClass) { it is Builder || it is Configuring || it is Adding || it is Restricted || it is HasDefaultValue } }
```

Comment:
💭 Should we extract and reuse the set of annotations? Given that the set of members that are callable via this mechanism should be consistent with what the schema builder produces, I would say that we should make sure that the `isPublicAndRestricted` member filter uses the same set of annotations instead of duplicating it.